ethereum/core/src/test/java/org/hyperledger/besu/ethereum/core/encoding/SetCodeTransactionDecoderTest.java

```diff
@@ -0,0 +1,129 @@
+
+import org.hyperledger.besu.crypto.SECPSignature;
+import org.hyperledger.besu.datatypes.Address;
+import org.hyperledger.besu.datatypes.SetCodeAuthorization;
+import org.hyperledger.besu.ethereum.rlp.BytesValueRLPInput;
+
+import java.math.BigInteger;
+
+import org.apache.tuweni.bytes.Bytes;
+import org.junit.jupiter.api.Test;
+
+class SetCodeTransactionDecoderTest {
+
+  @Test
+  void shouldDecodeInnerPayloadWithNonce() {
+    // "0xd80194633688abc3ccf8b0c03088d2d1c6ae4958c2fa56c105"
+
+    final BytesValueRLPInput input =
+        new BytesValueRLPInput(
+            Bytes.fromHexString(
+                "0xf85b0194633688abc3ccf8b0c03088d2d1c6ae4958c2fa56c18080a0840798fa67118e034c1eb7e42fe89e28d7cd5006dc813d5729e5f75b0d1a7ec5a03b1dbace38ceb862a65bf2eac0637693b5c3493bcb2a022dd614c0a74cce0b99"),
+            true);
+    final SetCodeAuthorization authorization = SetCodeTransactionDecoder.decodeInnerPayload(input);
+
+    assertThat(authorization.chainId()).isEqualTo(BigInteger.ONE);
+    assertThat(authorization.address())
+        .isEqualTo(Address.fromHexStringStrict("0x633688abc3cCf8B0C03088D2d1C6ae4958c2fA56"));
+    assertThat(authorization.nonces().size()).isEqualTo(1);
+    assertThat(authorization.nonces().getFirst()).isEqualTo(0L);
+
+    final SECPSignature signature = authorization.signature();
+    assertThat(signature.getRecId()).isEqualTo((byte) 0);
+    assertThat(signature.getR().toString(16))
+        .isEqualTo("840798fa67118e034c1eb7e42fe89e28d7cd5006dc813d5729e5f75b0d1a7ec5");
+    assertThat(signature.getS().toString(16))
+        .isEqualTo("3b1dbace38ceb862a65bf2eac0637693b5c3493bcb2a022dd614c0a74cce0b99");
+  }
+
+  @Test
+  void shouldDecodeInnerPayloadWithoutNonce() {
+    // "0xd70194633688abc3ccf8b0c03088d2d1c6ae4958c2fa56c5"
+
+    final BytesValueRLPInput input =
+        new BytesValueRLPInput(
+            Bytes.fromHexString(
+                "0xf85a0194633688abc3ccf8b0c03088d2d1c6ae4958c2fa56c001a0dd6b24048be1b7d7fe5bbbb73ffc37eb2ce1997ecb4ae5b6096532ef19363148a025b58a1ff8ad00bddbbfa1d5c2411961cbb6d08dcdc8ae88303db3c6cf983031"),
+            true);
+    final SetCodeAuthorization authorization = SetCodeTransactionDecoder.decodeInnerPayload(input);
+
+    assertThat(authorization.chainId()).isEqualTo(BigInteger.ONE);
+    assertThat(authorization.address())
+        .isEqualTo(Address.fromHexStringStrict("0x633688abc3cCf8B0C03088D2d1C6ae4958c2fA56"));
+    assertThat(authorization.nonces().size()).isEqualTo(0);
+
+    final SECPSignature signature = authorization.signature();
+    assertThat(signature.getRecId()).isEqualTo((byte) 1);
+    assertThat(signature.getR().toString(16))
+        .isEqualTo("dd6b24048be1b7d7fe5bbbb73ffc37eb2ce1997ecb4ae5b6096532ef19363148");
+    assertThat(signature.getS().toString(16))
+        .isEqualTo("25b58a1ff8ad00bddbbfa1d5c2411961cbb6d08dcdc8ae88303db3c6cf983031");
+  }
+
+  @Test
+  void shouldDecodeInnerPayloadWithMultipleNonces() {
```

Comment:
Invalid test per EIP-7702:

_When the list length is zero, consider the authorization nonce to be null. When the list length is one, consider the single integer value to be the provided nonce authorization. Other lengths and value types in this optional are invalid and the transaction as a whole should be considered malformed._